### PR TITLE
Move alternate calls to thrift hive metastore

### PR DESCRIPTION
**Pull request by skrzypo987**
## Description
See commit description

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftMetastoreModule.java
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftMetastoreModule.java:43]: We don't have a convention to import this statically.

```
$ git grep 'import com.google.inject.Scopes;' | wc -l
129

$ git grep 'import static com.google.inject.Scopes.SINGLETON;' | wc -l
22
```

let's not change it now.
you can offer a change in a separate PR

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/DefaultThriftMetastoreClientFactory.java
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/DefaultThriftMetastoreClientFactory.java]: move `metastoreSupportsDateStatistics` there too

Then change commit message to something like
```
Encapsulate metastore introspection state

The knowledge about supported metastore operations collected at runtime
is captured as a bunch of fields (mostly some `AtomicInteger` objects).
The commit introduces a new class to keep this information, so that its lifecycle
can be easier managed.
```

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreState.java
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreState.java:20]: UNDEFINED -> UNDEFINED_ALTERNATIVE ?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreState.java
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreState.java:30]: You can init the fields directly at their definition, no need for explicit ctor

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/DefaultThriftMetastoreClientFactory.java
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/DefaultThriftMetastoreClientFactory.java]: in the cmt msg

replace "...from ThriftMetastoreClient."
with "Move alternating calls logic to ThriftHiveMetastore from ThriftMetastoreClient."

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java:54]: > This partly reverts https://github.com/trinodb/trino/commit/280fee3038d49c1bc82015bdac29616b592d78ef

How much of the commit is the revert, and how much is new code?
I.e. what should i focus on in the review?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreState.java
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreState.java:18]: It doesn't describe "metastore state" (i read "state" as something that usually can change over time)
It describes "knowledge about metastore" or "metastore capabilities"

The "capabilities" is especially suitable for the metastoreSupportsDateStatistics field (to be moved to this class)
The "knowledge" fits better for the alternative calls...

Let's call this `ThriftHiveMetastoreIntrospection` 
and let's call fields/variables `metastoreIntrospection`.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java
Comment (skrzypo987) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java:54]: I removed the partly keyword

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreIntrospection.java
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreIntrospection.java:22]: Not sure why you moved ThriftMetastoreStats here
that wasn't my intention, sorry if i told you wrong

my intention was to move `private final MetastoreSupportsDateStatistics metastoreSupportsDateStatistics = new MetastoreSupportsDateStatistics();` here.

**Review by findepi — CHANGES_REQUESTED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreIntrospection.java
Comment (skrzypo987) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreIntrospection.java:22]: My bad, I mixed the names and blindly followed

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/DefaultThriftMetastoreClientFactory.java
Comment (dain) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/DefaultThriftMetastoreClientFactory.java:82]: There is no need to inject this.  The current code this replaces constructs `AtomicInteger` inline.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftMetastoreModule.java
Comment (dain) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftMetastoreModule.java:45]: remove this

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/FailureAwareThriftMetastoreClient.java
Comment (dain) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/FailureAwareThriftMetastoreClient.java:93]: This commit is not "Move alternate calls logic to ThriftHiveMetastore", but a massive refactor of the APIs for the ThriftMetastoreClient.  If we want to rename methods, those should be move to a separate commit.

**Review by dain — CHANGES_REQUESTED** [review]
I am ok with the first commit as it is cleaner, but we should not merge the second commit.  It makes the code much more complex and is not needed by any standard HMS.  I am aware that waggle dance has some broken logic it's Thrift implementation, but that should be fixed in the third party system.

**Comment by findepi** [comment]
> I am ok with the first commit as it is cleaner, but we should not merge the second commit. It makes the code much more complex and is not needed by any standard HMS. 

It's needed for standard HMS as well

I guess this wisdom comes with boring work on product tests and product tests environments, a work that only some maintainers have had their hands dirty with, so it's OK if you don't know that.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/FailureAwareThriftMetastoreClient.java
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/FailureAwareThriftMetastoreClient.java:93]: > This commit is not "Move alternate calls logic to ThriftHiveMetastore", but a massive refactor of the APIs for the ThriftMetastoreClient

that's mostly a revert of some previous refactor that turned out to be harmful
so the current "massive refactor" doesn't introduce any new quality that we haven't had before

Comment (skrzypo987) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/FailureAwareThriftMetastoreClient.java:93]: I couldn't do a simple revert, as these classes has been changed along the way. But yes, this is just a revert, no additional logic is implemented here.
IMHO following the single responsibility principle, the *Client class should just wrap the underlying thrift code, without any logic.
Another thing is that if, for any reason, the change that's being reverted here does add any value, we can simply put a config option that would decide whether the alternate calls should reuse the client.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/DefaultThriftMetastoreClientFactory.java
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/DefaultThriftMetastoreClientFactory.java]: this should be gone

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/FailureAwareThriftMetastoreClient.java
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/FailureAwareThriftMetastoreClient.java:93]: > put a config option that would decide whether the alternate calls should reuse the client.

there is no benefit to that, only downsides, so let's not do that

**Review by findepi — APPROVED** [review]

**Pull request closed without merging**
